app/controllers/sign_up/recovery_codes_controller.rb

```diff
@@ -27,18 +27,16 @@ def new_code
 
     def confirm_has_not_already_viewed_recovery_code
       return if user_session[:first_time_recovery_code_view].present?
-      redirect_to after_sign_in_path_for(current_user)
+      redirect_to after_sign_in_path_for
     end
 
     def next_step
-      if session[:saml_request_url]
+      if session[:sp]
         sign_up_completed_path
```

Comment:
I think there is an issue with our logic here (pre-dating this PR): if there is a service provider, is someone necessarily signing up? Is there ever a situation where they'd be referred by an sp but not signing up?